File: Album/src/main/java/gorden/album/fragment/AlbumPreViewFragment.java
package gorden.album.fragment;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.view.ViewPager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import java.util.ArrayList;

import gorden.album.AlbumPickerActivity;
import gorden.album.R;
import gorden.album.adapter.FragmentPagerAdapter;
import gorden.album.utils.AnimationUtils;
import gorden.album.widget.ZoomOutPageTransformer;

import static android.app.Activity.RESULT_OK;
import static gorden.album.AlbumPicker.EXTRA_MAX_COUNT;
import static gorden.album.AlbumPicker.EXTRA_PREVIEW_MODE;
import static gorden.album.AlbumPicker.EXTRA_SELECT_MODE;
import static gorden.album.AlbumPicker.KEY_DEL_IMAGES;
import static gorden.album.AlbumPicker.KEY_IMAGES;
import static gorden.album.fragment.AlbumPickerFragment.MULTI_SELECT_MODE;
import static gorden.album.fragment.AlbumPickerFragment.SINGLE_SELECT_MODE;

/**
 * 预览界面
 */

public class AlbumPreViewFragment extends Fragment implements View.OnClickListener {
    public static final String KEY_PREVIEW_POSITION = "KEY_PREVIEW_POSITION";
    public static final String KEY_PREVIEW_IMAGELIST = "KEY_PREVIEW_IMAGELIST";
    public static final String KEY_PREVIEW_SELECTED = "KEY_PREVIEW_SELECTED";

    public static final int MODE_ALBUM_PREVIEW = 701;//相册预览
    public static final int MODE_ALBUM_DELETE = 702;//图片列表删除
    public static final int MODE_ONLY_PREVIEW = 703;//仅仅预览

    private int previewMode = MODE_ALBUM_PREVIEW;

    private ViewPager pager_image;
    private TextView btn_confirm;
    private ImageButton btn_back;
    private TextView text_position;
    private CheckBox checkbox;
    private View viewClicked;

    private View rootView;
    private ArrayList<String> imgList = new ArrayList<>();
    private ArrayList<String> selectPath = new ArrayList<>();
    private int pickerMaxCount;
    private int currentPosition;
    private int pickerModel;

    private FragmentPagerAdapter pagerAdapter;

    private boolean show = true;
    private View rel_bottom, rel_top;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        if (rootView == null) {
            rootView = inflater.inflate(R.layout.fragment_album_preview1, container, false);
            pager_image = (ViewPager) rootView.findViewById(R.id.pager_image);
            btn_confirm = (TextView) rootView.findViewById(R.id.btn_confirm);
            btn_back = (ImageButton) rootView.findViewById(R.id.btn_back);
            text_position = (TextView) rootView.findViewById(R.id.text_position);
            checkbox = (CheckBox) rootView.findViewById(R.id.checkbox);
            viewClicked = rootView.findViewById(R.id.viewClicked);
            rel_top = rootView.findViewById(R.id.rel_top);
            rel_bottom = rootView.findViewById(R.id.rel_bottom);
        }
        return rootView;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        Bundle arguments = getArguments();

        previewMode = arguments.getInt(EXTRA_PREVIEW_MODE);
        currentPosition = getArguments().getInt(KEY_PREVIEW_POSITION, 0);
        imgList = getArguments().getStringArrayList(KEY_PREVIEW_IMAGELIST);

        //保证数据正常
        if (imgList == null || imgList.size() == 0) return;
        if (currentPosition > imgList.size()-1) currentPosition = 0;

        pager_image.setPageTransformer(false, new ZoomOutPageTransformer());
        pager_image.setAdapter(pagerAdapter = new FragmentPagerAdapter(getChildFragmentManager(), imgList));
        pager_image.setCurrentItem(currentPosition);
        ((AlbumPickerActivity) getActivity()).applyBackground(imgList.get(currentPosition));
        text_position.setText((currentPosition + 1) + "/" + imgList.size());

        switch (previewMode) {
            case MODE_ALBUM_PREVIEW://相册
                ArrayList<String> tempSelect = getArguments().getStringArrayList(KEY_PREVIEW_SELECTED);
                if (tempSelect != null) selectPath.addAll(tempSelect);

                pickerMaxCount = getArguments().getInt(EXTRA_MAX_COUNT);
                pickerModel = getArguments().getInt(EXTRA_SELECT_MODE);

                refreshConfirm();
                checkbox.setChecked(selectPath.contains(imgList.get(currentPosition)));
                break;
            case MODE_ALBUM_DELETE://图片选择删除
                btn_confirm.setText("删除");
                btn_confirm.setEnabled(true);
                rel_bottom.setVisibility(View.GONE);
                break;
            case MODE_ONLY_PREVIEW://仅仅预览
                btn_confirm.setVisibility(View.GONE);
                rel_bottom.setVisibility(View.GONE);
                break;
        }

        pager_image.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

            }

            @Override
            public void onPageSelected(int position) {
                currentPosition = position;
                text_position.setText((position + 1) + "/" + imgList.size());
                ((AlbumPickerActivity) getActivity()).applyBackground(imgList.get(position));
                if (pickerModel == MULTI_SELECT_MODE)
                    checkbox.setChecked(selectPath.contains(imgList.get(position)));
            }

            @Override
            public void onPageScrollStateChanged(int state) {

            }
        });

        if (previewMode == MODE_ALBUM_PREVIEW || previewMode == MODE_ALBUM_DELETE) {
            btn_confirm.setOnClickListener(this);
        }

        if (previewMode == MODE_ALBUM_PREVIEW && pickerModel == MULTI_SELECT_MODE) {
            viewClicked.setOnClickListener(this);
        }

        btn_back.setOnClickListener(this);
    }

    /**
     * 刷新完成按钮
     */
    public void refreshConfirm() {
        if (pickerModel == SINGLE_SELECT_MODE) {
            btn_confirm.setText("确定");
            btn_confirm.setEnabled(true);
            rel_bottom.setVisibility(View.GONE);
        } else {
            btn_confirm.setText(selectPath.size() > 0 && pickerMaxCount > 1 ? "完成(" + selectPath.size() + "/" + pickerMaxCount + ")" : "完成");
            btn_confirm.setEnabled(selectPath.size() > 0);
        }
    }

    public void onBackPressed(){
        if (getFragmentManager().getBackStackEntryCount() > 0)
            getFragmentManager().popBackStack();
        else{
            if (previewMode == MODE_ALBUM_DELETE && selectPath.size()>0){
                Intent resultData = new Intent();
                resultData.putStringArrayListExtra(KEY_DEL_IMAGES, selectPath);
                getActivity().setResult(RESULT_OK, resultData);
            }
            getActivity().finish();
        }
    }

    public void toggleToolbarVisibleState() {
        show = !show;
        if (show) {
            AnimationUtils.visibleViewByAlpha(rel_top);
            if (pickerModel != SINGLE_SELECT_MODE && previewMode == MODE_ALBUM_PREVIEW)
                AnimationUtils.visibleViewByAlpha(rel_bottom);
        } else {
            AnimationUtils.goneViewByAlpha(rel_top);
            AnimationUtils.goneViewByAlpha(rel_bottom);
        }
    }

    @Override
    public void onClick(View v) {
        if (v.getId() == R.id.btn_back) {//返回按钮
            onBackPressed();
        } else if (v.getId() == R.id.btn_confirm) {//确认
            if (previewMode == MODE_ALBUM_DELETE){//编辑删除模式
                if (imgList.size()>0){
                    selectPath.add(imgList.get(currentPosition));
                    imgList.remove(currentPosition);
                    pagerAdapter.notifyDataSetChanged();
                    text_position.setText((currentPosition + 1) + "/" + imgList.size());
                }
                if (imgList.size()==0){
                    text_position.setText("0/0");
                    Intent resultData = new Intent();
                    resultData.putStringArrayListExtra(KEY_DEL_IMAGES, selectPath);
                    getActivity().setResult(RESULT_OK, resultData);
                    getActivity().finish();
                }
            }else{//相册选择模式
                Intent resultData = new Intent();
                //单选模式,返回当前图片
                if (pickerModel == SINGLE_SELECT_MODE) {
                    selectPath.clear();
                    selectPath.add(imgList.get(currentPosition));
                }
                resultData.putStringArrayListExtra(KEY_IMAGES, selectPath);
                getActivity().setResult(RESULT_OK, resultData);
                getActivity().finish();
            }
        } else if (v.getId() == R.id.viewClicked) {
            boolean isChecked = !checkbox.isChecked();
            if (isChecked && selectPath.size() >= pickerMaxCount) {
                Toast.makeText(getActivity(), "您最多只能选择" + pickerMaxCount + "张照片", Toast.LENGTH_SHORT).show();
                return;
            }
            checkbox.setChecked(isChecked);

            if (isChecked && !selectPath.contains(imgList.get(currentPosition))) {
                selectPath.add(imgList.get(currentPosition));
            } else if (!isChecked && selectPath.contains(imgList.get(currentPosition))) {
                selectPath.remove(imgList.get(currentPosition));
            }
            ((AlbumPickerActivity) getActivity()).refreshSelected(selectPath);
            refreshConfirm();
        }
    }
}

File: lib_album/src/main/kotlin/gorden/library/ui/AlbumPreviewFragment.kt
package gorden.library.ui

import android.annotation.SuppressLint
import android.app.Activity.RESULT_OK
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentStatePagerAdapter
import android.support.v4.view.PagerAdapter
import android.support.v4.view.ViewPager
import android.support.v7.app.AlertDialog
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.SimpleItemAnimator
import android.view.LayoutInflater
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.ViewGroup
import android.widget.Toast
import gorden.library.Album.Companion.KEY_DEL_IMAGES
import gorden.library.Album.Companion.KEY_IMAGES
import gorden.library.R
import gorden.library.aac.PictureViewModel
import gorden.library.adapter.SelectedAdapter
import gorden.library.core.ZoomOutPageTransformer
import gorden.library.entity.*
import kotlinx.android.synthetic.main.fragment_album_preview.*


/**
 * 图片预览
 */
class AlbumPreviewFragment : Fragment() {
    private lateinit var pictureViewModel: PictureViewModel
    private lateinit var adapterSelected: SelectedAdapter
    val currentPosition = MutableLiveData<Int>()
    private var scroller: Scroller? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_album_preview, container, false)
    }

    @SuppressLint("SetTextI18n")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        pictureViewModel = ViewModelProviders.of(activity!!).get(PictureViewModel::class.java)

        currentPosition.value = pictureViewModel.previewPosition

        try {
            val field = ViewPager::class.java.getDeclaredField("mScroller")
            field.isAccessible = true
            scroller = Scroller(context)
            field.set(pager_image, scroller)
        } catch (e: NoSuchFieldException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        }


        var previewPaths: List<String> = pictureViewModel.selectedPaths
        when (pictureViewModel.previewMode) {
            PREVIEW_ALBUM -> {
                btn_confirm.visibility = VISIBLE
                previewPaths = pictureViewModel.currentDirectory.value?.pictures!!.map { it.path }
                adapterSelected = SelectedAdapter(pictureViewModel.selectedPaths, pictureViewModel, this)
                recycler_selected.layoutManager = LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
                (recycler_selected.itemAnimator as SimpleItemAnimator).supportsChangeAnimations = false
                recycler_selected.adapter = adapterSelected
            }
            PREVIEW_SELECT -> {
                btn_confirm.visibility = VISIBLE
                previewPaths = pictureViewModel.selectedPaths.toList()
                adapterSelected = SelectedAdapter(pictureViewModel.selectedPaths.toList(), pictureViewModel, this)
                recycler_selected.layoutManager = LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
                (recycler_selected.itemAnimator as SimpleItemAnimator).supportsChangeAnimations = false
                recycler_selected.adapter = adapterSelected
            }
            PREVIEW_PREVIEW -> {
                lin_bottom.visibility = GONE
            }
            PREVIEW_DELETE -> {
                btn_delete.visibility = VISIBLE
                lin_bottom.visibility = GONE
            }
        }

        currentPosition.observe(this, Observer {
            it?.apply {
                when (pictureViewModel.previewMode) {
                    PREVIEW_ALBUM -> {
                        checkbox.isChecked = pictureViewModel.selectedPaths.contains(pictureViewModel.currentDirectory.value!!.pictures[this].path)
                    }
                    PREVIEW_SELECT -> {
                        checkbox.isChecked = pictureViewModel.selectedPaths.contains(adapterSelected.paths[this])
                    }
                }
                text_position.text = "${this + 1}/${previewPaths.size}"
                scroller?.interception = true
                pager_image.setCurrentItem(this, true)
                scroller?.interception = false
            }
        })

        refreshConfirm()

        pager_image.setPageTransformer(false, ZoomOutPageTransformer())

        pager_image.adapter = object : FragmentStatePagerAdapter(childFragmentManager) {
            override fun getItem(position: Int): Fragment {
                return AlbumImageFragment().apply {
                    arguments = Bundle().apply { putString("url", previewPaths[position]) }
                }
            }

            override fun getCount(): Int {
                return previewPaths.size
            }

            override fun getItemPosition(`object`: Any): Int {
                return PagerAdapter.POSITION_NONE
            }
        }

        pager_image.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            override fun onPageScrollStateChanged(state: Int) {

            }

            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
            }

            override fun onPageSelected(position: Int) {
                currentPosition.value = position
            }
        })

        checkbox.setOnCheckedChangeListener { compoundButton, checked ->
            if (compoundButton.isPressed) {//手动选择
                if (checked) {
                    if (pictureViewModel.selectedPaths.size < pictureViewModel.maxCount) {
                        if (pictureViewModel.previewMode == PREVIEW_SELECT) {
                            pictureViewModel.selectedPaths.add(previewPaths[currentPosition.value!!])
                            adapterSelected.notifyItemChanged(currentPosition.value!!)
                        } else {
                            pictureViewModel.selectedPaths.add(previewPaths[currentPosition.value!!])
                            if (pictureViewModel.selectedPaths.size == 1) {//刷新闪烁的问题
                                adapterSelected.notifyDataSetChanged()
                            } else {
                                adapterSelected.notifyItemInserted(pictureViewModel.selectedPaths.size - 1)
                            }
                            recycler_selected.scrollToPosition(pictureViewModel.selectedPaths.size - 1)
                        }
                        pictureViewModel.notifySelected()
                        refreshConfirm()
                    } else {
                        compoundButton.isChecked = false
                        Toast.makeText(activity, "您最多只能选择" + pictureViewModel.maxCount + "张照片", Toast.LENGTH_SHORT).show()
                    }
                } else {
                    if (pictureViewModel.previewMode == PREVIEW_SELECT) {
                        pictureViewModel.selectedPaths.remove(previewPaths[currentPosition.value!!])
                        adapterSelected.notifyItemChanged(currentPosition.value!!)
                    } else {
                        val index = pictureViewModel.selectedPaths.indexOf(previewPaths[currentPosition.value!!])
                        pictureViewModel.selectedPaths.removeAt(index)
                        adapterSelected.notifyItemRemoved(index)
                    }
                    pictureViewModel.notifySelected()
                    refreshConfirm()
                }
            }
        }


        btn_back.setOnClickListener {
            activity?.onBackPressed()
        }

        btn_confirm.setOnClickListener {
            val resultData = Intent()
            if (pictureViewModel.selectedPaths.isEmpty()) {
                resultData.putStringArrayListExtra(KEY_IMAGES, arrayListOf(previewPaths[currentPosition.value!!]))
            } else {
                resultData.putStringArrayListExtra(KEY_IMAGES, pictureViewModel.selectedPaths)
            }
            activity?.setResult(RESULT_OK, resultData)
            activity?.finish()
        }

        btn_delete.setOnClickListener {
            AlertDialog.Builder(activity!!).setMessage("要删除这张照片吗?")
                    .setPositiveButton("确定", { _, _ ->
                        pictureViewModel.delList.add((previewPaths as ArrayList).removeAt(currentPosition.value!!))
                        if (previewPaths.isEmpty()) {
                            val resultData = Intent()
                            resultData.putStringArrayListExtra(KEY_DEL_IMAGES, pictureViewModel.delList)
                            activity?.setResult(RESULT_OK, resultData)
                            activity?.onBackPressed()
                        } else {
                            pager_image.adapter?.notifyDataSetChanged()
                            text_position.text = "${currentPosition.value!! + 1}/${previewPaths.size}"
                        }
                    })
                    .setNegativeButton("取消", null)
                    .show().apply {
                        getButton(AlertDialog.BUTTON_POSITIVE).setTextColor(Color.BLACK)
                        getButton(AlertDialog.BUTTON_NEGATIVE).setTextColor(Color.LTGRAY)
                    }
        }
    }

    private fun refreshConfirm() {
        when (pictureViewModel.selectMode) {
            SINGLE_SELECT_MODE -> {
//                checkbox.visibility = GONE
//                view_line.visibility = GONE
//                recycler_selected.visibility = GONE
                lin_bottom.visibility = GONE
            }
            MULTI_SELECT_MODE -> {
                val size = pictureViewModel.selectedPaths.size
                btn_confirm.text = if (size > 0 && pictureViewModel.maxCount > 1)
                    String.format(getString(R.string.album_str_complete), "($size/${pictureViewModel.maxCount})")
                else
                    String.format(getString(R.string.album_str_complete), "")
                if (size > 0) {
                    view_line.visibility = VISIBLE
                    recycler_selected.visibility = VISIBLE
                } else {
                    view_line.visibility = GONE
                    recycler_selected.visibility = GONE
                }
            }
        }
    }

    private class Scroller(context: Context?) : android.widget.Scroller(context) {
        var interception: Boolean = false
        override fun startScroll(startX: Int, startY: Int, dx: Int, dy: Int, duration: Int) {
            super.startScroll(startX, startY, dx, dy, if (interception) 0 else duration)
        }
    }
}